### PR TITLE
Project Rosetta

### PR DESCRIPTION
**Pull request by colegottdank**

### CODE REVIEW

File: web/lib/wrappers/rosetta/rosettaWrapper.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { RosettaCache } from "./rosettaCache";
+
+export class RosettaWrapper {
+  private rosetta: Rosetta;
+  private rosettaStore: IDatabaseService;
+  private rosettaCache: ICacheService;
+  constructor(supabaseClient: SupabaseClient<Database>) {
+    this.rosettaStore = new RosettaStore(supabaseClient);
+    this.rosettaCache = new RosettaCache();
+    this.rosetta = new Rosetta({
```
Comment (colegottdank) [inline web/lib/wrappers/rosetta/rosettaWrapper.ts:20]: Need to update these env variables

File: web/yarn.lock [hunk 2]
```diff
@@ -3115,6 +3124,14 @@
   resolved "https://registry.yarnpkg.com/@types/lru-cache/-/lru-cache-5.1.1.tgz#c48c2e27b65d2a153b19bfc1a317e30872e01eef"
   integrity sha512-ssE3Vlrys7sdIzs5LOxCzTVMsU7i9oa/IaW92wF32JFb3CVczqOkru2xspuKczHEbG3nvmPY7IFqVmGGHdNbYw==
 
+"@types/node-fetch@^2.6.4":
```
Comment (colegottdank) [inline web/yarn.lock:3127]: Why so many changes in yarn lock?

File: worker/src/index.ts [hunk 3]
```diff
@@ -137,8 +140,18 @@ export default {
     env: Env,
     ctx: ExecutionContext
   ): Promise<void> {
-    await feedbackCronHandler(env);
-    await updateLoopUsers(env);
+    switch (controller.cron) {
```
Comment (colegottdank) [inline worker/src/index.ts:143]: Ensure I add this cron job to all environments

File: worker/src/lib/rosetta/RosettaStore.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    const { error } = await this.supabaseClient
+      .from("rosetta_mappers")
+      .insert(newMapper);
+
+    if (error) {
+      console.error("Failed to add rosetta mapper", error);
+    }
+  }
+
+  private ToDbMapper(
```
Comment (colegottdank) [inline worker/src/lib/rosetta/RosettaStore.ts:68]: ToDbMapper and ToRosettaMapper should be abstracted into the Rosetta package

File: worker/src/lib/rosetta/RosettaWrapper.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import { RosettaStore } from "./RosettaStore";
+import { Env } from "../..";
+
+export class RosettaWrapper {
+  private rosetta: Rosetta;
+  private rosettaStore: IDatabaseService;
+
+  constructor(supabaseClient: SupabaseClient<Database>, env: Env) {
+    this.rosettaStore = new RosettaStore(supabaseClient);
+    this.rosetta = new Rosetta({
```
Comment (colegottdank) [inline worker/src/lib/rosetta/RosettaWrapper.ts:14]: Ensure this environment variables get set!

File: supabase/migrations/20231011172312_rosetta_init.sql [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    'PENDING_APPROVAL',
+    'ACTIVE',
+    'INACTIVE',
+    'DECLINED',
+    'FAILED'
+);
+CREATE TABLE rosetta_mappers (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    key TEXT NOT NULL,
+    -- A composite key (e.g., "OpenAI:GPT3.5")
```
Comment (chitalian) [inline supabase/migrations/20231011172312_rosetta_init.sql:14]: Is this still true? Or are we moving to path?

File: web/components/templates/jobs/jobsPage.tsx [hunk 7]
```diff
@@ -15,7 +15,6 @@ import { clsx } from "../../shared/clsx";
 import ThemedTableV5 from "../../shared/themed/table/themedTableV5";
 import { ThemedSwitch } from "../../shared/themed/themedSwitch";
 import { NormalizedRequest } from "../requestsV2/builder/abstractRequestBuilder";
-import getRequestBuilder from "../requestsV2/builder/requestBuilder";
```
Comment (chitalian) [inline web/components/templates/jobs/jobsPage.tsx:18]: thanks

File: web/components/templates/requestsV2/builder/chatBuilder.tsx
Comment (chitalian) [inline web/components/templates/requestsV2/builder/chatBuilder.tsx:1]: I am so happy we are doing this...

File: web/lib/wrappers/rosetta/rosettaWrapper.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  ): Promise<Json | null> {
+    const outputSchema = JSON.parse(JSON.stringify(requestResponseSchema));
+    const key = `${provider}:${requestPath}`;
+
+    try {
+      return await this.rosetta.map(requestResponse, outputSchema, key);
+    } catch (error: any) {
+      console.log(`Error mapping LLM call: ${error.message}`);
+      return null;
+    }
```
Comment (chitalian) [inline web/lib/wrappers/rosetta/rosettaWrapper.ts:49]: COLE THIS IS AWESOME!

File: web/lib/wrappers/rosetta/rosettaWrapper.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  Json,
+  Rosetta,
+} from "@helicone/project-rosetta";
+import requestResponseSchema from "./requestResponseSchema.json";
+import { RosettaStore } from "./rosettaStore";
+import { SupabaseClient } from "@supabase/supabase-js";
+import { Database } from "../../../supabase/database.types";
+import { RosettaCache } from "./rosettaCache";
+
+export class RosettaWrapper {
```
Comment (chitalian) [inline web/lib/wrappers/rosetta/rosettaWrapper.ts:13]: lol - when you build a service and then wrap it

File: web/lib/wrappers/rosetta/rosettaStore.ts [hunk 10]
```diff
@@ -0,0 +1,108 @@
+import { SupabaseClient } from "@supabase/supabase-js";
+import { Database } from "../../../supabase/database.types";
+import {
+  IDatabaseService,
+  RosettaMapper,
+  RosettaMapperStatus,
+} from "@helicone/project-rosetta";
+
+export class RosettaStore implements IDatabaseService {
```
Comment (chitalian) [inline web/lib/wrappers/rosetta/rosettaStore.ts:9]: this is sooo coool

File: web/lib/wrappers/rosetta/requestResponseSchema.json
Comment (chitalian) [inline web/lib/wrappers/rosetta/requestResponseSchema.json:1]: fun fact this is only 1.3k tokens

File: web/lib/wrappers/rosetta/rosettaCache.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import {
+  ICacheService,
+  RosettaMapper,
+} from "@helicone/project-rosetta/dist/src/RosettaTypes";
+import { kv } from "@vercel/kv";
```
Comment (chitalian) [inline web/lib/wrappers/rosetta/rosettaCache.ts:5]: Did you make the KV in prod?

File: web/lib/wrappers/rosetta/rosettaWrapper.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    try {
+      return await this.rosetta.map(requestResponse, outputSchema, key);
+    } catch (error: any) {
+      console.log(`Error mapping LLM call: ${error.message}`);
+      return null;
+    }
+  }
+
+  public async generateMappers() {
+    await this.rosetta.generateMappers();
```
Comment (chitalian) [inline web/lib/wrappers/rosetta/rosettaWrapper.ts:53]: I F**ing love composition.

File: web/pages/api/request/index.ts [hunk 13]
```diff
@@ -24,10 +25,17 @@ async function handler({
     sort: SortLeafRequest;
     isCached: boolean;
   };
-  const metrics = isCached
-    ? await getRequestsCached(orgId, filter, offset, limit, sort)
-    : await getRequests(orgId, filter, offset, limit, sort);
 
+  const metrics = isCached
```
Comment (chitalian) [inline web/pages/api/request/index.ts:29]: ew so much more lines of code lol

File: worker/src/lib/rosetta/RosettaWrapper.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import { IDatabaseService } from "@helicone/project-rosetta/dist/src/RosettaTypes";
+import { SupabaseClient } from "@supabase/supabase-js";
+import { Database } from "../../../supabase/database.types";
+import { Rosetta } from "@helicone/project-rosetta/dist/src/Rosetta";
+import { RosettaStore } from "./RosettaStore";
+import { Env } from "../..";
+
+export class RosettaWrapper {
```
Comment (chitalian) [inline worker/src/lib/rosetta/RosettaWrapper.ts:8]: Wait I am confused- why do we need the mapper in the worker too?

File: worker/src/index.ts [hunk 15]
```diff
@@ -137,8 +140,18 @@ export default {
...
+      case "0 * * * *":
+        const supabaseClient = createClient(
+          env.SUPABASE_URL,
+          env.SUPABASE_SERVICE_ROLE_KEY
+        );
+        const rosetta = new RosettaWrapper(supabaseClient, env);
+        await rosetta.generateMappers();
```
Comment (chitalian) [inline worker/src/index.ts:150]: break?

File: worker/src/index.ts (hunk 3 above)
Comment (chitalian) [inline worker/src/index.ts:143]: dude I hate switch statements lol

File: supabase/migrations/20231011172312_rosetta_init.sql (hunk 6 above)
Comment (colegottdank) [reply supabase/migrations/20231011172312_rosetta_init.sql:14]: Not true anymore, comment isn’t necessary

File: web/lib/wrappers/rosetta/rosettaWrapper.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        apiKey: "",
+        organization: "",
+      },
+      heliconeApiKey: "",
+      requireApproval: true,
+    });
+  }
+
+  public async mapLLMCall(
+    requestResponse: {
```
Comment (colegottdank) [inline web/lib/wrappers/rosetta/rosettaWrapper.ts:33]: LlmCall*

File: worker/src/lib/rosetta/RosettaWrapper.ts (hunk 14 above)
Comment (colegottdank) [reply worker/src/lib/rosetta/RosettaWrapper.ts:8]: This is just for the Cron job. V2 I think we should have 2 different types. RosettaMapper and RosettaGenerator

File: worker/src/index.ts (hunk 3 above)
Comment (colegottdank) [reply worker/src/index.ts:143]: I love them lol

Comment (colegottdank) [reply worker/src/index.ts:143]: It will only run in prod ^^

File: web/lib/wrappers/rosetta/rosettaCache.ts (hunk 11 above)
Comment (colegottdank) [reply web/lib/wrappers/rosetta/rosettaCache.ts:5]: yes

**Pull request merged**